lib/gollum/public/gollum/javascript/editor/modes.js.erb

```diff
@@ -0,0 +1,151 @@
+
+    * GitHub style code blocks: ``` ... ```
+    * UML blocks: @startuml ... @enduml
+
+    The regexes defined for these below should be equivalent to the regexes used by
+    gollum-lib to extract these blocks.
+
+    ACE Docs:
+      * Highlight Rules: https://cloud9-sdk.readme.io/docs/highlighting-rules
+      * Dynamically generating modes: https://stackoverflow.com/questions/22166784/dynamically-update-syntax-highlighting-mode-rules-for-the-ace-editor
+*/
+
+<%
+codeblock_rule_name = 'GitHubBlock'
+codeblock_rule_id = 'githubblock'
+%>
+
+// The rule defining the github-codeblock state for the ACE Tokenizer
+var <%= codeblock_rule_name %> = [{
+  token: "support.function",
+  regex: /^```$/,
+  next:  "start"
+}];
+
+// The rule defining how the ACE Tokenizer gets into the github-codeblock state
+var GitHubStart = {
+  token: "support.function",
+  regex: /^```/,
+  next:  "<%= codeblock_rule_id %>"
+};
+
+<%
+uml_rule_name = 'UMLBlock'
+uml_rule_id = 'umlblock'
+%>
+
+// The rule defining the UML state for the ACE Tokenizer
+var <%= uml_rule_name %>  = [{
+  token: "support.function",
+  regex: /^\@enduml/,
+  next:  "start"
+}];
+
+// The rule defining how the ACE Tokenizer gets into the UML state
+var UMLStart = {
+  token: "support.function",
+  regex: /^\@startuml/,
+  next:  "<%= uml_rule_id %>"
+};
+
+<%
+default_rules = {
+  :path => '',
+  :rule_name => '',
+  :start_rules => ['GitHubStart', 'UMLStart'],
+  :block_rules => {codeblock_rule_id => codeblock_rule_name, uml_rule_id => uml_rule_name}
+}
+
+rst_rules = default_rules.dup
+rst_rules[:path] = 'ace/mode/rst_highlight_rules'
+rst_rules[:rule_name] = 'RSTHighlightRules'
+
+asciidoc_rules = default_rules.dup
+asciidoc_rules[:path] = 'ace/mode/asciidoc_highlight_rules'
+asciidoc_rules[:rule_name] = 'AsciidocHighlightRules'
+
+textile_rules = default_rules.dup
+textile_rules[:path] = 'ace/mode/textile_highlight_rules'
+textile_rules[:rule_name] = 'TextileHighlightRules'
+
+{
+  'GollumRstHighlightRules'      => rst_rules,
+  'GollumAsciidocHighlightRules' => asciidoc_rules,
+  'GollumTextileHighlightRules'  => textile_rules,
+  'GollumMarkdownHighlightRules' => {
+    :path => 'ace/mode/markdown_highlight_rules',
+    :rule_name => 'MarkdownHighlightRules',
+    :start_rules => ['UMLStart'], # Markdown only needs to be extended with UML rules
+    :block_rules => {uml_rule_id => uml_rule_name}    
+  }
+}.each do |name, highlighter|
+%>
+// Register gollum-specific HighlightRules with ACE
+ace.define("<%= name %>", [], function(require, exports, module) {
+  "use strict";
+
+  var oop = require("ace/lib/oop");
+  var BaseHighlightRules = require('<%= highlighter[:path] %>').<%= highlighter[:rule_name] %>;
+
+  var GollumHighlightRules = function() {
+    var mergeRules = new BaseHighlightRules();
+    this.$rules = mergeRules.getRules();
+    <% highlighter[:start_rules].each do |rule| %>
+    this.$rules.start.unshift(<%= rule %>);
```

Comment:
Indentation missing.

Reply:
Think I prefer to stick to the identation level for the JS here